File: 백준/Silver/1463. 1로 만들기/1로 만들기.py
x=int(input()) # 수 입력받기
d=[0]*(x+1) # 1-based
for i in range(2,x+1): # 2부터 x까지 반복
    d[i]=d[i-1]+1 # 1을 빼는 연산 -> 1회 진행
    if i%2==0: # 2로 나누어 떨어질 때, 2로 나누는 연산
        d[i]=min(d[i],d[i//2]+1)
    if i%3==0: # 3으로 나누어 떨어질 때, 3으로 나누는 연산
        d[i]=min(d[i],d[i//3]+1)
print(d[x])

File: 백준/Silver/1463. 1로 만들기/1로 만들기.js
const readline = require("readline");

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let inputLine = 0;
rl.on("line", (line) => {
  if (inputLine == 0) {
    let N = parseInt(line);
    const dp = new Array(N + 1);
    dp[1] = 0;
    dp[2] = 1;
    dp[3] = Math.min(dp[1] + 1, dp[2] + 1);
    for (let i = 4; i <= N; i++) {
      if (i % 2 == 0) {
        dp[i] = Math.min(dp[i - 1] + 1, dp[i / 2] + 1);
      }
      if (i % 3 == 0) {
        dp[i] = Math.min(dp[i - 1] + 1, dp[i / 3] + 1);
      }
      if (i % 3 == 0 && i % 2 == 0) {
        dp[i] = Math.min(dp[i - 1] + 1, dp[i / 2] + 1, dp[i / 3] + 1);
      }
      if (i % 3 !== 0 && i % 2 !== 0) {
        dp[i] = dp[i - 1] + 1;
      }
    }
    console.log(dp[N]);
    process.exit();
  }
});
